tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -326,6 +314,51 @@ LogicalResult lowerWithScheduleLoop(
   return success();
 }
 
+LogicalResult lowerWithScheduleLoopV2(
+    ArrayRef<Operation*> root_ops, Operation* dominant_op,
+    Block* parent = nullptr, const ShapeAnalysis* shape_analysis = nullptr,
+    int vector_size = 1) {
+  const auto loc = dominant_op->getLoc();
+  OpBuilder b(root_ops.back());
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value vec_size = b.create<arith::ConstantIndexOp>(loc, vector_size);
+  auto num_elements = emitNumElementsComputation(b, loc, dominant_op);
+  auto thread_number = b.create<arith::DivUIOp>(loc, num_elements, vec_size);
+  Value var;
+  SmallVector<Value, 2> vars;
+  scf::ParallelOp simt_loop = createParallelAndSetInsPt(
```

Comment:
newbie question: is it here the `ParallelOp` where the vectorization magic happens?

Reply:
Not the ParallelOp. The ParallelOp is to form the SIMT parallelism structure (i.e. grid and block) of GPU kernel.
What helps to do the vectorization is the ForOp at line-336 bellow. After unroll and interleaving for this op, the instructions inside this ForOp could be vectorized after some LLVM passes.

Reply:
Thanks.